xdgopenproxy/xdgopenproxy.go

```diff
@@ -27,37 +27,95 @@ import (
 	"github.com/godbus/dbus"
 )
 
-func Run(urlOrFile string) error {
-	bus, err := dbus.SessionBus()
+type bus interface {
+	Object(name string, objectPath dbus.ObjectPath) dbus.BusObject
+	Close() error
+}
+
+var sessionBus = func() (bus, error) { return dbus.SessionBus() }
+
+type desktopLauncher interface {
+	openFile(path string) error
+	openURL(url string) error
+}
+
+type portalLauncher struct {
+	service dbus.BusObject
+}
+
+func (p *portalLauncher) openFile(filename string) error {
+	fd, err := syscall.Open(filename, syscall.O_RDONLY, 0)
```

Comment:
We should try to open the file descriptor with `O_RDWR` first, and fall back to `O_RDONLY` on error.  If we pass an `O_RDONLY` and the associated application is confined, then the portal will only give that application read access to the file.

Reply:
There's also some information about using `writable` key in extra arguments to `OpenFile()` here https://flatpak.github.io/xdg-desktop-portal/portal-docs.html#gdbus-org.freedesktop.portal.OpenURI Does opening the file in RDWR mode enough, or do we need to pass `writable: true` still?

Reply:
Hmm.  Looking at the GLib code, it seems they don't ever try to use the writable=true mode for OpenFile:

https://gitlab.gnome.org/GNOME/glib/blob/master/gio/gopenuriportal.c#L99

So maybe it isn't worth trying to pass an `O_RDWR` file descriptor.  That would make things a bit simpler.